### ISSUE TITLE
NaN loss on multi-gpu training

### ISSUE THREAD

**Issue by datasciantle** [tensorflow/tensorflow#87432]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

binary

### TensorFlow version

v1.12.1-122039-g784fed5357b 2.20.0-dev20250211

### Custom code

Yes

### OS platform and distribution

Amazon Linux 2

### Python version

3.10.14

### CUDA/cuDNN version

CUDA 12.3 (Build V12.3.107)

### GPU model and memory

GPU Model: NVIDIA A10G Memory per GPU: 24 GB

### Current behavior?

I find that training a distributed model - with either MirroredStrategy or MultiWorkerMirroredStrategy - that the loss jumps to NaN on the 61st batch.

The problem vanishes if I train on a single GPU (no distribute strategy), or if I use tf 2.13.1.
The problem also vanishes if I set the number of steps per epoch to be less than 61.

I've attached some reproducible code but note that I find the exact same finding (the role of batch 61, single GPU, v.2.13) on other datasets, other model architectures, and with different batch sizes.

### Standalone code to reproduce the issue

```shell
call this test.py
# these pass
python3 test.py --steps_per_epoch 60 --strategy None
python3 test.py --steps_per_epoch 60 --strategy MirroredStrategy
python3 test.py --steps_per_epoch 60 --strategy MultiWorkerMirroredStrategy
python3 test.py --steps_per_epoch 61 --strategy None

# these fail
python3 test.py --steps_per_epoch 61 --strategy MirroredStrategy
python3 test.py --steps_per_epoch 61 --strategy MultiWorkerMirroredStrategy



import tensorflow as tf
import numpy as np
import argparse

parser = argparse.ArgumentParser()
parser.add_argument('--batch_size', type=int, default=64)
parser.add_argument('--steps_per_epoch', type=int, default=128)
parser.add_argument('--strategy', type=str, default='None')
args = parser.parse_args()


# For reproducibility
tf.random.set_seed(42)
np.random.seed(42)

# Parameters
n_string = 256      # used for the CLS token value
n_features = 256   # length of the output label vector
dim_model = 64

batch_size = int(args.batch_size)
steps_per_epoch = int(args.steps_per_epoch)


def random_input(n_features, n_string):
    """
    Generate random input features in the integer domain.
    """
    return tf.random.uniform(shape=(n_features,), minval=0, maxval=n_string, dtype=tf.int32)

def random_mask(input_features, batch_masking_rate):
    """
    Randomly mask input features in the integer domain.
    Each element is retained with probability (1 - batch_masking_rate) and zeroed otherwise.
    """
    # Generate random values for each element in the same shape as input_features.
    rand_vals = tf.random.uniform(shape=tf.shape(input_features), minval=0.0, maxval=1.0)
    # Create a binary mask: retain element if random value >= masking_rate, else 0.
    # This yields 1 with probability (1 - batch_masking_rate) and 0 with probability batch_masking_rate.
    mask = tf.cast(rand_vals >= batch_masking_rate, dtype=tf.int32)
    return input_features * mask

def process_example(output_label):
    """
    Given an output_label vector, generate a random masking rate,
    apply random masking to produce input_features, and prepend a CLS token.
    All operations remain in the integer domain.
    """
    # Draw a random masking rate uniformly between 0 and 0.5.
    # (Note: masking_rate is used only for the masking decision and is not passed on.)
    masking_rate = tf.random.uniform([], minval=0.0, maxval=0.5)
    # output_label is already an integer tensor.
    masked_input = random_mask(output_label, masking_rate)
    # Return a features dictionary and a labels dictionary.
    features = {"input_features": masked_input}
    labels = {"the_label": output_label}
    return features, labels

def make_dataset(num_examples, n_features, n_string, batch_size):
    """
    Create a tf.data.Dataset of random output labels.
    Each output label is a random integer tensor of shape (n_features,).
    """
    # Create a dataset of num_examples random output labels.
    dummy_examples = [random_input(n_features, n_string) for _ in range(num_examples)]
    # Build a tf.data.Dataset from the dummy output labels.
    dataset = tf.data.Dataset.from_tensor_slices(dummy_examples)
    dataset = dataset.map(process_example)
    dataset = dataset.batch(batch_size)
    return dataset.repeat()

def make_the_model(n_features, n_string, dim_model):
    """
    Build a simple model that takes input_features and predicts the_label.
    """
    input_features = tf.keras.layers.Input(shape=(n_features,), dtype=tf.int32, name="input_features")
    embedding = tf.keras.layers.Embedding(n_string, dim_model)(input_features)
    mha = tf.keras.layers.MultiHeadAttention(
        num_heads=8,
        key_dim=dim_model//8,
        dropout=0.,
        name='mha_0'
    )(embedding, embedding)
    mha = tf.keras.layers.MultiHeadAttention(
        num_heads=8,
        key_dim=dim_model//8,
        dropout=0.,
        name='mha_1'
    )(mha, mha)
    # Flatten the embedding and apply a dense layer to predict the_label.
    the_label = tf.keras.layers.Dense(n_features, activation=None, name="the_label")(mha)
    model = tf.keras.Model(inputs={'input_features': input_features}, outputs={'the_label': the_label})
    optimizer = tf.keras.optimizers.Adam(learning_rate=0.001)
    loss_dict = {'the_label': tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)}
    model.compile(optimizer=optimizer, loss=loss_dict)
    return model

training_data = make_dataset(1024, n_features, n_string, batch_size)
validation_data = make_dataset(256, n_features, n_string, batch_size)

# print an example of the training data
# for batch in training_data.take(1):
#     print(batch)

# Build the model
if args.strategy == 'MirroredStrategy':
    print("Using MirroredStrategy.")
    strategy = tf.distribute.MirroredStrategy()
    with strategy.scope():
        this_model = make_the_model(n_features, n_string, dim_model)
        this_model.summary()
        print(f'Training with batch size {batch_size} and steps per epoch {steps_per_epoch} with {strategy.num_replicas_in_sync} replicas.')


if args.strategy == 'MultiWorkerMirroredStrategy':
    print("Using MultiWorkerMirroredStrategy.")
    strategy = tf.distribute.MultiWorkerMirroredStrategy()
    with strategy.scope():
        this_model = make_the_model(n_features, n_string, dim_model)
        this_model.summary()
        print(f'Training with batch size {batch_size} and steps per epoch {steps_per_epoch} with {strategy.num_replicas_in_sync} replicas.')

if args.strategy == 'None':
    print("No strategy.")
    this_model = make_the_model(n_features, n_string, dim_model)
    this_model.summary()
    print(f'Training with batch size {batch_size} and steps per epoch {steps_per_epoch} with 1 GPU.')



# Train the model
call_backs = tf.keras.callbacks.TerminateOnNaN()
this_model.fit(training_data, validation_data=validation_data, epochs=100, steps_per_epoch=steps_per_epoch, validation_steps=4, callbacks=[call_backs])```

### Relevant log output

```shell

```

**Comment by Venkat6871** [comment]
Hi **@datasciantle** ,
Apologies for the delay, and thanks for raising your concern here. I tried running your code using TensorFlow 2.18.0 (stable) on VM instances, but I did not encounter any issues. I am attaching the output below for your reference. And I recommend using stable versions for better results.
```
[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU'), PhysicalDevice(name='/physical_device:GPU:1', device_type='GPU')]
```
```
(tf_env) maayara@venkat-gpu1:~$ python3 test.py --steps_per_epoch 61 --strategy MultiWorkerMirroredStrategy
2025-03-05 11:01:28.725763: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
E0000 00:00:1741172488.746097   33345 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1741172488.752394   33345 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2025-03-05 11:01:28.774422: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
I0000 00:00:1741172491.290716   33345 gpu_device.cc:2022] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 13764 MB memory:  -> device: 0, name: Tesla T4, pci bus id: 0000:00:04.0, compute capability: 7.5
I0000 00:00:1741172491.293431   33345 gpu_device.cc:2022] Created device /job:localhost/replica:0/task:0/device:GPU:1 with 13764 MB memory:  -> device: 1, name: Tesla T4, pci bus id: 0000:00:05.0, compute capability: 7.5
Using MultiWorkerMirroredStrategy.
Model: "functional"
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                  ┃ Output Shape              ┃         Param # ┃ Connected to               ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input_features (InputLayer)   │ (None, 256)               │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ embedding (Embedding)         │ (None, 256, 64)           │          16,384 │ input_features[0][0]       │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ mha_0 (MultiHeadAttention)    │ (None, 256, 64)           │          16,640 │ embedding[0][0],           │
│                               │                           │                 │ embedding[0][0]            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ mha_1 (MultiHeadAttention)    │ (None, 256, 64)           │          16,640 │ mha_0[0][0], mha_0[0][0]   │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ the_label (Dense)             │ (None, 256, 256)          │          16,640 │ mha_1[0][0]                │
└───────────────────────────────┴───────────────────────────┴─────────────────┴────────────────────────────┘
 Total params: 66,304 (259.00 KB)
 Trainable params: 66,304 (259.00 KB)
 Non-trainable params: 0 (0.00 B)
Training with batch size 64 and steps per epoch 61 with 2 replicas.
Epoch 1/100
61/61 ━━━━━━━━━━━━━━━━━━━━ 10s 48ms/step - loss: 5.5452 - val_loss: 5.5454
Epoch 2/100
61/61 ━━━━━━━━━━━━━━━━━━━━ 2s 38ms/step - loss: 5.5448 - val_loss: 5.5454
.
.
.
.
Epoch 97/100
61/61 ━━━━━━━━━━━━━━━━━━━━ 2s 40ms/step - loss: 1.3885 - val_loss: 1.3389
Epoch 98/100
61/61 ━━━━━━━━━━━━━━━━━━━━ 2s 38ms/step - loss: 1.4202 - val_loss: 1.3394
Epoch 99/100
61/61 ━━━━━━━━━━━━━━━━━━━━ 3s 44ms/step - loss: 1.3504 - val_loss: 1.3385
Epoch 100/100
61/61 ━━━━━━━━━━━━━━━━━━━━ 3s 44ms/step - loss: 1.4391 - val_loss: 1.3389

Thank you!
